File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.*;
import java.util.stream.*;

public class Solution {
    private static Map<String, Integer> playerInfo = new HashMap<>();

    public String[] solution(String[] players, String[] callings) {
        makeInfo(players);

        Arrays.stream(callings).forEach( player -> {
            int idx = playerInfo.get(player);
            String frontPlayerName = players[idx - 1];
            players[idx] = frontPlayerName;
            players[idx - 1] = player;

            playerInfo.put(frontPlayerName, idx);
            playerInfo.put(player, idx - 1);
        });

        return players;
    }

    private static void makeInfo(String[] players) {
        IntStream.range(0, players.length).forEach(index -> {
            playerInfo.put(players[index], index);
        });
    }
}

File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.js
var playerInfo = {};

function solution(players, callings) {
    makeInfo(players);

    callings.forEach(call => {
        let idx = playerInfo[call];
        let frontPlayerName = players[idx - 1];

        players[idx] = players[idx - 1];
        players[idx - 1] = call;

        playerInfo[call] = idx - 1;
        playerInfo[frontPlayerName] = idx;

    });
    return players;
}

function makeInfo(players) {
    players.forEach( (player, index) => {
        playerInfo[player] = index;
    })
}